### PR TITLE
Fix Remote Code Execution vulnerability

### LINKED ISSUE #1359: Remote Code Execution Vulnerability
# Report

## What did you do?

Run `danger` on a malicious branch name.

## What did you expect to happen?

No such malicious commands are executed.

## What happened instead?

A malicious command run.

## Your Environment

* Which CI are you running on? **GitHub Actions**
* Are you running the latest version of Danger? **Yes**
* What is your Dangerfile? **Default from `danger init`**

## Why

First, `git_fetch_branch_to_depth` in `Danger::GitRepo` executes `Danger::GitRepo#exec` with parameter of a branch name.

https://github.com/danger/danger/blob/v8.5.0/lib/danger/scm_source/git_repo.rb#L109

Then, `Danger::GitRepo#exec` executes `Open3.popen2` with the branch name.

https://github.com/danger/danger/blob/v8.5.0/lib/danger/scm_source/git_repo.rb#L58

And finally, it lead to remote code execution vulneralibty if the branch name is malicious.

Branch names are limited, but sufficient to execute commands. (Ref: [git check-ref-format --branch](https://git-scm.com/docs/git-check-ref-format))

## Other

I have reported this vulnerability to @orta in advance and got his permission.

### PR DESCRIPTION
**Pull request by kyoshidajp**
Fixes #1359

My idea is `Danger::GitRepo#exec` (`Danger::LocalGitRepo#run_git`, `Danger::LocalOnlyGitRepo#run_git`) executes `Open3.popen2` with expanding Array parameter.

### CODE REVIEW

File: spec/lib/danger/scm_source/git_repo_spec.rb [hunk 1]
```diff
@@ -4,12 +4,10 @@
   describe "#exec" do
     it "run command with our env set" do
       git_repo = described_class.new
-      allow(git_repo).to receive(:default_env) { Hash("LANG" => "zh_TW.UTF-8") }
-      command = Gem.win_platform? ? "status && set LANG" : "status && echo $LANG"
+      allow(git_repo).to receive(:default_env) { Hash("GIT_EXEC_PATH" => "git-command-path") }
+      result = git_repo.exec(["--exec-path"])
```
Comment (kyoshidajp) [inline spec/lib/danger/scm_source/git_repo_spec.rb:8]: Control operators (e.g. `&&`, `||`) are not available.

File: spec/lib/danger/scm_source/git_repo_spec.rb [hunk 2]
```diff
@@ -4,12 +4,10 @@
   describe "#exec" do
     it "run command with our env set" do
       git_repo = described_class.new
-      allow(git_repo).to receive(:default_env) { Hash("LANG" => "zh_TW.UTF-8") }
-      command = Gem.win_platform? ? "status && set LANG" : "status && echo $LANG"
+      allow(git_repo).to receive(:default_env) { Hash("GIT_EXEC_PATH" => "git-command-path") }
```
Comment (kyoshidajp) [inline spec/lib/danger/scm_source/git_repo_spec.rb:7]: [`GIT_EXEC_PATH`](https://git-scm.com/book/en/v2/Git-Internals-Environment-Variables) is useful in this case.

`git --exec-path` can check the value.

File: spec/lib/danger/scm_source/git_repo_spec.rb [hunk 3]
```diff
@@ -4,12 +4,10 @@
   describe "#exec" do
     it "run command with our env set" do
       git_repo = described_class.new
-      allow(git_repo).to receive(:default_env) { Hash("LANG" => "zh_TW.UTF-8") }
-      command = Gem.win_platform? ? "status && set LANG" : "status && echo $LANG"
```
Comment (kyoshidajp) [inline spec/lib/danger/scm_source/git_repo_spec.rb:8]: Control operators (e.g. `&&`, `||`) are not available.

**Comment by orta** [comment]
Looks great, thanks for finding this!

**Comment by orta** [comment]
Shipped in Danger 8.6.0

**Pull request merged**